concordium-std-derive/src/lib.rs

```diff
@@ -33,47 +37,275 @@ fn attach_error<A>(mut v: syn::Result<A>, msg: &str) -> syn::Result<A> {
-/// return an Err. If the item does not exist the return value is Ok(None).
-/// FIXME: Ensure there is only one.
-fn get_attribute_value<'a, I: IntoIterator<Item = &'a Meta>>(
-    iter: I,
-    name: &str,
-) -> syn::Result<Option<&'a syn::LitStr>> {
+/// Attributes that can be attached either to the init or receive method of a
+/// smart contract.
+struct OptionalArguments {
+    /// If set, the contract can receive gTU.
+    pub(crate) payable:       bool,
+    /// If enabled, the function has access to logging facilities.
+    pub(crate) enable_logger: bool,
+    /// The function is a low-level one, with direct access to contract memory.
+    pub(crate) low_level:     bool,
+    /// Which type, if any, is the parameter type of the contract.
+    /// This is used when generating schemas.
+    pub(crate) parameter:     Option<syn::LitStr>,
+}
+
+/// Attributes that can be attached to the initialization method.
+struct InitAttributes {
+    /// Name of the contract.
+    pub(crate) contract: syn::LitStr,
+    pub(crate) optional: OptionalArguments,
+}
+
+/// Attributes that can be attached to the receive method.
+struct ReceiveAttributes {
+    /// Name of the contract the method applies to.
+    pub(crate) contract: syn::LitStr,
+    /// Name of the method.
+    pub(crate) name:     syn::LitStr,
+    pub(crate) optional: OptionalArguments,
+}
+
+#[derive(Default)]
+struct ParsedAttributes {
+    /// We use BTreeSet to have consistent order of iteration when reporting
+    /// errors.
+    pub(crate) flags:  BTreeSet<syn::Ident>,
+    /// We use BTreeMap to have consistent order of iteration when reporting
+    /// errors.
+    pub(crate) values: BTreeMap<syn::Ident, syn::LitStr>,
+}
+
+impl ParsedAttributes {
+    /// Remove an attribute and return it, if present. The key must be a valid
+    /// Rust identifier, otherwise this function will panic.
+    pub(crate) fn extract_value(&mut self, key: &str) -> Option<syn::LitStr> {
+        // This is not clean, constructing a new identifier with a call_site span.
+        // But the only alternative I see is iterating over the map and locating the key
+        // since Ident implements equality comparison with &str.
+        let key = syn::Ident::new(key, Span::call_site());
+        self.values.remove(&key)
+    }
+
+    /// Remove an attribute return whether it was present.
+    pub(crate) fn extract_flag(&mut self, key: &str) -> bool {
+        // This is not clean, constructing a new identifier with a call_site span.
+        // But the only alternative I see is iterating over the map and locating the key
+        // since Ident implements equality comparison with &str.
+        let key = syn::Ident::new(key, Span::call_site());
+        self.flags.remove(&key)
+    }
+
+    /// If there are any remaining attributes signal an error. Otherwise return
+    /// Ok(())
+    pub(crate) fn report_all_attributes(self) -> syn::Result<()> {
```

Comment:
Consider renaming it to `report_unrecognized_attributes` (or similar) to clarify its purpose.

Reply:
That was in fact the original name of this function :) but then I changed it since the current name is more descriptive.

The `report_unrecognized_attributes` makes sense in the context this is used.

So I would prefer to keep it.